lib/iris/analysis/_grid_angles.py

```diff
@@ -0,0 +1,390 @@
+# (C) British Crown Copyright 2010 - 2018, Met Office
+#
+# This file is part of Iris.
+#
+# Iris is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Iris is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Code to implement vector rotation by angles, and inferring gridcell angles
+from coordinate points and bounds.
+
+"""
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+import numpy as np
+
+import iris
+
+
+def _3d_xyz_from_latlon(lon, lat):
+    """
+    Return locations of (lon, lat) in 3D space.
+
+    Args:
+
+    * lon, lat: (arrays in degrees)
+
+    Returns:
+
+        xyz : (array, dtype=float64)
+            cartesian coordinates on a unit sphere.  Dimension 0 maps x,y,z.
+
+    """
+    lon1 = np.deg2rad(lon).astype(np.float64)
+    lat1 = np.deg2rad(lat).astype(np.float64)
+
+    x = np.cos(lat1) * np.cos(lon1)
+    y = np.cos(lat1) * np.sin(lon1)
+    z = np.sin(lat1)
+
+    result = np.concatenate([array[np.newaxis] for array in (x, y, z)])
+
+    return result
+
+
+def _latlon_from_xyz(xyz):
+    """
+    Return arrays of lons+lats angles from xyz locations.
+
+    Args:
+
+    * xyz: (array)
+        positions array, of dims (3, <others>), where index 0 maps x/y/z.
+
+    Returns:
+
+        lonlat : (array)
+            spherical angles, of dims (2, <others>), in radians.
+            Dim 0 maps longitude, latitude.
+
+    """
+    lons = np.arctan2(xyz[1], xyz[0])
+    axial_radii = np.sqrt(xyz[0] * xyz[0] + xyz[1] * xyz[1])
+    lats = np.arctan2(xyz[2], axial_radii)
+    return np.array([lons, lats])
+
+
+def _angle(p, q, r):
+    """
+    Return angle (in _radians_) of grid wrt local east.
+    Anticlockwise +ve, as usual.
+    {P, Q, R} are consecutive points in the same row,
+    eg {v(i,j),f(i,j),v(i+1,j)}, or {T(i-1,j),T(i,j),T(i+1,j)}
+    Calculate dot product of PR with lambda_hat at Q.
+    This gives us cos(required angle).
+    Disciminate between +/- angles by comparing latitudes of P and R.
+    p, q, r, are all 2-element arrays [lon, lat] of angles in degrees.
+
+    """
+#    old_style = True
+    old_style = False
+    if old_style:
```

Comment:
How do you plan to determine whether or not old_style should be used?

Reply:
Sorry, this was an alternative calculation that I used for evaluation.
I have since removed it (cf. https://github.com/pp-mo/iris/pull/45)

Though... I accept that I ***still*** don't have a full understanding of this calculation :grimacing: .
We should probably make more effort to explain exactly what this is doing.
Put that also down to #3109 !